### ISSUE TITLE
Warn about dangerous uses of unsafeRun

### ISSUE THREAD

**Issue by reibitto** [zio/zio#2201]
This was brought up on Gitter by discussion between cchepelov and @neko-kai. https://gitter.im/ZIO/Core?at=5dc00ae08fbe1d5faba1fc5e

I primarily wanted to make an issue about it so that it doesn't potentially get lost.

Not sure if my understanding is 100% correct, but it seems like the sync unsafe methods are more dangerous than the async variants because it's easier to cause deadlocks such as when you run it on the thread pool that ZIO is using.

Is it possible to detect this kind of usage and warn (or error out?) to prevent people from running into it and wondering what's going on?

Or if that's not really feasible, at the very least add a usage warning to the scaladoc? I think it would be useful to know which unsafe methods are more dangerous than the others and why.

**Comment by neko-kai** [comment]
Now we have `Fiber._currentFiber` ThreadLocal. We can fail all unsafeRuns for which it's not `null`.

**Comment by IgorDorokhov** [comment]
Hi everyone, is this issue still relevant?
As per the description of the unsafeRunSync() function in Cats library the function is blocking the thread indefinitely.

Snippet from the description:
"If any component of the computation is asynchronous, the current thread will block awaiting the results of the async computation. By default, this blocking will be unbounded. "

Function itself (from package cats.effect):
```
  final def unsafeRunSync()(implicit runtime: unsafe.IORuntime): A =
    unsafeRunTimed(Long.MaxValue.nanos).get
```
Which means it causes a deadlock under certain circumstances. For example, if we call unsafeRunSync within a Fiber.
Like so:
```
for {
fiber <- effect.fork
_ <- ZIO.runtime[Any].flatMap{ runtime => runtime.unsafeRunAsyc(fiber.join)(_ => ())
} yield ()
```
In the scenario above fiber.join cannot be scheduled because the thread is blocked, hence we get a deadlock.

To avoid that kind of situation a following check can be added `Fiber._currentFiber.isDefined`. In case true is returned an exception should be thrown (for example,     throw new IllegalStateException("unsafeRun cannot be called from within an executing effect as it may cause deadlocks.")

Is my understanding correct?

Since I am new to ZIO I need some guidance on where the check should be added.
Thank you.

**Comment by IgorDorokhov** [comment]
Oh, I see. unsafeRunSync was deprecated in ZIO2, it is available in ZIO1 only.
Should this issue be closed then?

**Comment by ghostdogpr** [comment]
Yeah now we even require an implicit `Unsafe` to run it, so I think it's enough?

**Comment by kyri-petrou** [comment]
> Oh, I see. unsafeRunSync was deprecated in ZIO2, it is available in ZIO1 only.
> Should this issue be closed then?

@IgorDorokhov in ZIO 2.x there is the equivalent `Runtime#unsafe.run` method. The good news is that the behaviour that you're describing (that was leading to deadlocks previously) was fixed in v2.1.5. Check #8955 for more info on the issue and the fix.

I think we can close this issue for now and reopen if you think the current behaviour might still lead to deadlocks.
